### PR TITLE
Extend RuntimeHelpers.IsBitwiseEquatable to more types

### PR DESCRIPTION
**Pull request by stephentoub**
Today, RuntimeHelpers.IsBitwiseEquatable is hardcoded to a fixed list of types.  This means that almost all of the vectorization we've done with arrays and spans is limited to just those types; developers can themselves use MemoryMarshal.Cast to convert spans of other types to spans of supported one, but it doesn't naturally happen.

This extends IsBitwiseEquatable a bit more. We already have a CanCompareBitsOrUseFastGetHashCode helper used by ValueType.Equals to determine whether structs that don't override Equals can be compared with the equivalent of memcmp.  This extends that same helper to be used by IsBitwiseEquatable.  However, IsBitwiseEquatable also needs to rule out types that implement `IEquatable<T>` (the existing helper doesn't because it's about the implementation of the object.Equals override where the interface doesn't come into play).

The upside of this is APIs like Array.IndexOf will now automatically vectorize with more types (and avoid all the boxing overhead and the like with object.Equals).  The main downside is that types which provide their own equality implementation still don't benefit, which in turn means adding an `IEquality<T>` implementation could in the future be a deoptimization (we should consider some kind of attribute or marker interface a type can use to say "I promise my equality implementation is the same as a bitwise comparison"); plus, anyone who realizes they have a perf issue will have already added the interface implementation.  We also currently constrain most of our MemoryExtensions methods to types that implement `IEquatable<T>`, so there are only a handful of public methods today that benefit from this.

|             Method |         Toolchain |         Mean | Ratio | Allocated | Alloc Ratio |
|------------------- |------------------ |-------------:|------:|----------:|------------:|
|            IndexOf | \main\corerun.exe | 13,711.25 ns | 1.000 |   48000 B |        1.00 |
|            IndexOf |   \pr\corerun.exe |     68.16 ns | 0.005 |         - |        0.00 |
|                    |                   |              |       |           |             |
| CommonPrefixLength | \main\corerun.exe | 13,005.48 ns | 1.000 |   48000 B |        1.00 |
| CommonPrefixLength |   \pr\corerun.exe |    129.64 ns | 0.010 |         - |        0.00 |
|                    |                   |              |       |           |             |
|     SequenceEquals | \main\corerun.exe | 13,342.51 ns | 1.000 |   48000 B |        1.00 |
|     SequenceEquals |   \pr\corerun.exe |     70.01 ns | 0.005 |         - |        0.00 |

```C#
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using System;
using System.Linq;

[MemoryDiagnoser(displayGenColumns: false)]
[HideColumns("Job", "Error", "StdDev")]
public partial class Program
{
    static void Main(string[] args) => BenchmarkSwitcher.FromAssembly(typeof(Program).Assembly).Run(args);

    private MyColor[] _values1, _values2;

    [GlobalSetup]
    public void Setup()
    {
        _values1 = Enumerable.Range(0, 1_000).Select(i => new MyColor { R = (byte)i, G = (byte)i, B = (byte)i, A = (byte)i }).ToArray();
        _values2 = (MyColor[])_values1.Clone();
    }

    [Benchmark]
    public int IndexOf() => Array.IndexOf(_values1, new MyColor { R = 1, G = 2, B = 3, A = 4 });

    [Benchmark]
    public int CommonPrefixLength() => _values1.AsSpan().CommonPrefixLength(_values2);

    [Benchmark]
    public bool SequenceEquals() => _values1.AsSpan().SequenceEqual(_values2);
}

struct MyColor
{
    public byte R, G, B, A;
}
```

### CODE REVIEW

File: src/coreclr/vm/jitinterface.cpp
Comment (jkotas) [inline src/coreclr/vm/jitinterface.cpp:7236]: Could you please make a matching change for the AOT compilers?

https://github.com/dotnet/runtime/blob/ab2d195f169a08bf8eb978b472b2d3aca61b0a46/src/coreclr/tools/Common/TypeSystem/IL/Stubs/RuntimeHelpersIntrinsics.cs#L54

File: src/coreclr/vm/jitinterface.cpp
Comment (MichalPetryka) [inline src/coreclr/vm/jitinterface.cpp:7152]: Does this properly account for padding between struct members?

File: src/coreclr/vm/jitinterface.cpp
Comment (MichalPetryka) [inline src/coreclr/vm/jitinterface.cpp:7147]: I don't see a reason to block this for types with unmanaged pointers?

Comment (jkotas) [reply src/coreclr/vm/jitinterface.cpp:7147]: This change does not block it for unmanaged pointers.

`MethodTable::ContainsPointers` returns true for managed object references only.

File: src/coreclr/vm/jitinterface.cpp
Comment (EgorBo) [inline src/coreclr/vm/jitinterface.cpp:7158]: Is an overriden `Equals` in a struct sign that it should not be used for bitwise equality?

File: src/coreclr/vm/jitinterface.cpp
Comment (EgorBo) [reply src/coreclr/vm/jitinterface.cpp:7152]: Yes, it also handles floating point fields (rejects in case of them)

File: src/coreclr/vm/jitinterface.cpp
Comment (stephentoub) [reply src/coreclr/vm/jitinterface.cpp:7158]: Isn't that handled by https://github.com/dotnet/runtime/blob/53841055e735895520633ecf5da392ac14415ba8/src/coreclr/vm/comutilnative.cpp#L1670?

Comment (EgorBo) [reply src/coreclr/vm/jitinterface.cpp:7158]: @stephentoub thanks, it's interesting that that one doesn't take `IEquatable<T>` into account 🙂 so I assume fast hashcode/fast equals on structs might ignore it

File: src/coreclr/vm/jitinterface.cpp
Comment (MichalStrehovsky) [reply src/coreclr/vm/jitinterface.cpp:7236]: We have the algorithm in https://github.com/dotnet/runtime/blob/fa42038b64236a49b0a74776ec9a8b321b95b98c/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/CompilerTypeSystemContext.ValueTypeMethods.cs#L116 but it's currently not callable from RuntimeHelpersIntrinsics.cs.

I would pull that out into a separate static method in a separate file (without all the caching) and change that place and this place to call the static method.

We probably don't need this to be behind a cache for IsBitwiseEquatable.

File: src/coreclr/vm/jitinterface.cpp
Comment (stephentoub) [reply src/coreclr/vm/jitinterface.cpp:7158]: I assume that one doesn't because it doesn't need to: it's being used from ValueType.Equals/GetHashCode, so it was only being invoked in a situation where `IEquatable<T>` isn't relevant because it wouldn't have been used anyway (plus `IEquatable<T>` was added much later).

**Comment by stephentoub** [comment]
Ugh, fun:
```
Assert failure(PID 4956 [0x0000135c], Thread: 5712 [0x1650]): CONTRACT VIOLATION by CanCompareBitsOrUseFastGetHashCode at "D:\a\_work\1\s\src\coreclr\vm\comutilnative.cpp" @ 1649

MODE_COOPERATIVE encountered while thread is in preemptive state.

                        CONTRACT in CanCompareBitsOrUseFastGetHashCode at "D:\a\_work\1\s\src\coreclr\vm\comutilnative.cpp" @ 1649
VIOLATED-->  CONTRACT in getILIntrinsicImplementationForRuntimeHelpers at "D:\a\_work\1\s\src\coreclr\vm\jitinterface.cpp" @ 7169
                        CONTRACT in getMethodInfoHelper at "D:\a\_work\1\s\src\coreclr\vm\jitinterface.cpp" @ 7435
                        CONTRACT in UnsafeJitFunction at "D:\a\_work\1\s\src\coreclr\vm\jitinterface.cpp" @ 12716
                        CONTRACT in MethodDesc::JitCompileCodeLocked at "D:\a\_work\1\s\src\coreclr\vm\prestub.cpp" @ 935
                        CONTRACT in MethodDesc::JitCompileCodeLockedEventWrapper at "D:\a\_work\1\s\src\coreclr\vm\prestub.cpp" @ 770
                        CONTRACT in MethodDesc::JitCompileCode at "D:\a\_work\1\s\src\coreclr\vm\prestub.cpp" @ 650
                        CONTRACT in MethodDesc::PrepareILBasedCode at "D:\a\_work\1\s\src\coreclr\vm\prestub.cpp" @ 348
                        CONTRACT in MethodDesc::PrepareCode at "D:\a\_work\1\s\src\coreclr\vm\prestub.cpp" @ 318
                        CONTRACT in CodeVersionManager::PublishVersionableCodeIfNecessary at "D:\a\_work\1\s\src\coreclr\vm\codeversion.cpp" @ 1639
                        CONTRACT in MethodDesc::DoPrestub at "D:\a\_work\1\s\src\coreclr\vm\prestub.cpp" @ 2014
                        GCX_PREEMP_THREAD_EXISTS in PreStubWorker at "D:\a\_work\1\s\src\coreclr\vm\prestub.cpp" @ 1936
                        CONTRACT in DispatchCallSimple at "D:\a\_work\1\s\src\coreclr\vm\callhelpers.cpp" @ 178
                        CONTRACT in ThreadNative::KickOffThread_Worker at "D:\a\_work\1\s\src\coreclr\vm\comsynchronizable.cpp" @ 148
                        CONTRACT in ManagedThreadBase_DispatchInner at "D:\a\_work\1\s\src\coreclr\vm\threads.cpp" @ 7245
                        CONTRACT in ManagedThreadBase_FullTransition at "D:\a\_work\1\s\src\coreclr\vm\threads.cpp" @ 7493
                        CONTRACT in ThreadNative::KickOffThread at "D:\a\_work\1\s\src\coreclr\vm\comsynchronizable.cpp" @ 196
```

File: src/coreclr/vm/jitinterface.cpp
Comment (stephentoub) [reply src/coreclr/vm/jitinterface.cpp:7236]: Thanks. Will do.

**Comment by jkotas** [comment]
> MODE_COOPERATIVE 

Changing the contract on `CanCompareBitsOrUseFastGetHashCode` from `MODE_COOPERATIVE` to `MODE_ANY` should fix this.

File: src/coreclr/vm/jitinterface.cpp
Comment (tannergooding) [inline src/coreclr/vm/jitinterface.cpp:7235]: Is this problematic for that case (is it Mono only?) where an enum can be backed by a `float` or `double`?

Comment (stephentoub) [reply src/coreclr/vm/jitinterface.cpp:7235]: https://github.com/dotnet/runtime/blob/ad8debf62d9ae538c4e1a2f473a5a86377eabbc8/src/libraries/System.Private.CoreLib/src/System/Enum.cs#L12-L22

File: src/coreclr/vm/jitinterface.cpp
Comment (stephentoub) [reply src/coreclr/vm/jitinterface.cpp:7236]: I made some changes locally, but haven't been able to successfully build (sent you a note in Teams).  In the meantime I pushed up the changes just to see how CI feels about it and to get your eyes on it in case I misinterpreted your feedback.

File: src/coreclr/tools/Common/TypeSystem/IL/Stubs/RuntimeHelpersIntrinsics.cs
Comment (jkotas) [inline src/coreclr/tools/Common/TypeSystem/IL/Stubs/RuntimeHelpersIntrinsics.cs:86]:
```suggestion
```
Left-over from Char8 experiment

File: src/coreclr/tools/Common/TypeSystem/IL/Stubs/RuntimeHelpersIntrinsics.cs
Comment (jkotas) [inline src/coreclr/tools/Common/TypeSystem/IL/Stubs/RuntimeHelpersIntrinsics.cs:84]: The previous condition suggests that mdType can be null.

File: src/coreclr/tools/Common/TypeSystem/IL/Stubs/RuntimeHelpersIntrinsics.cs
Comment (jkotas) [inline src/coreclr/tools/Common/TypeSystem/IL/Stubs/RuntimeHelpersIntrinsics.cs:80]:
```suggestion
                        if (mdType != null && mdType.Module == mdType.TypeSystemContext.SystemModule && mdType.Name == "Rune" && mdType.Namespace == "System.Text")
```
Unrelated bug (we do not want to pick user types that happen to have this name)

**Review by jkotas — APPROVED** [review]
LGTM modulo comments

File: src/coreclr/tools/Common/TypeSystem/IL/Stubs/RuntimeHelpersIntrinsics.cs
Comment (jkotas) [inline src/coreclr/tools/Common/TypeSystem/IL/Stubs/RuntimeHelpersIntrinsics.cs:83]:
```suggestion
                                mdType.Name == "Rune")
```

Comment (stephentoub) [reply src/coreclr/tools/Common/TypeSystem/IL/Stubs/RuntimeHelpersIntrinsics.cs:83]: Ugh. Thanks. I need to get this compiling locally; still having trouble even with main.

File: src/coreclr/tools/Common/TypeSystem/IL/Stubs/ComparerIntrinsics.cs
Comment (MichalStrehovsky) [inline src/coreclr/tools/Common/TypeSystem/IL/Stubs/ComparerIntrinsics.cs:5]: Nit: we separate System namespaces from internal namespaces with a blank line. It's intentional.

File: src/coreclr/tools/Common/TypeSystem/IL/Stubs/ComparerIntrinsics.cs
Comment (stephentoub) [inline src/coreclr/tools/Common/TypeSystem/IL/Stubs/ComparerIntrinsics.cs:5]:
```suggestion

using Internal.TypeSystem;
```

**Pull request merged**
